bad error message for wrong type of typeddict setitem

<!--
  If you're new to mypy and you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form
  instead.
-->

**Bug Report**

<!--
Note: If the problem you are reporting is about a specific library function, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues
-->

confusing error message

**To Reproduce**

```
from typing import TypedDict

Foo = TypedDict('Foo', {'a': int})
blah: Foo = {'a': 1}
blah['a'] = 'lol'
```

**Expected Behavior**

something like "value of 'a' should be `int`, not `str`"

<!--
  How did you expect your project to behave?
  It’s fine if you’re not sure your understanding is correct.
  Write down what you thought would happen. If you just expected no errors, you can delete this section.
-->


**Actual Behavior**

```
a.py:5: error: Argument 2 has incompatible type "str"; expected "int"
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 0.812
- Python version used: Python 3.8.0

<!--
You can freely edit this text, please remove all the lines
you believe are unnecessary.
-->

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -548,7 +548,7 @@ class MessageBuilder:
                     expected_type,
                     bare=True)
                 arg_label = '"{}"'.format(arg_name)
-            if isinstance(outer_context, IndexExpr) and isinstance(outer_context.index, StrExpr):
+            if isinstance(outer_context, IndexExpr):
                 msg = 'Argument "{}" has incompatible type {}; expected {}' .format(
                     outer_context.index.value, quote_type_string(arg_type_str),
                     quote_type_string(expected_type_str))

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -548,7 +548,7 @@ class MessageBuilder:
                     expected_type,
                     bare=True)
                 arg_label = '"{}"'.format(arg_name)
-            if isinstance(outer_context, IndexExpr):
+            if isinstance(outer_context, IndexExpr) and isinstance(outer_context.index, StrExpr):
                 msg = 'Argument "{}" has incompatible type {}; expected {}' .format(
                     outer_context.index.value, quote_type_string(arg_type_str),
                     quote_type_string(expected_type_str))

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -31,7 +31,7 @@ from mypy.nodes import (
     FuncDef, reverse_builtin_aliases,
     ARG_POS, ARG_OPT, ARG_NAMED, ARG_NAMED_OPT, ARG_STAR, ARG_STAR2,
     ReturnStmt, NameExpr, Var, CONTRAVARIANT, COVARIANT, SymbolNode,
-    CallExpr, SymbolTable, TempNode
+    CallExpr, IndexExpr, StrExpr, SymbolTable, TempNode
 )
 from mypy.subtypes import (
     is_subtype, find_member, get_member_flags,
@@ -536,7 +536,6 @@ class MessageBuilder:
                 arg_name = outer_context.arg_names[n - 1]
                 if arg_name is not None:
                     arg_label = '"{}"'.format(arg_name)
-
             if (arg_kind == ARG_STAR2
                     and isinstance(arg_type, TypedDictType)
                     and m <= len(callee.arg_names)
@@ -549,9 +548,14 @@ class MessageBuilder:
                     expected_type,
                     bare=True)
                 arg_label = '"{}"'.format(arg_name)
-            msg = 'Argument {} {}has incompatible type {}; expected {}'.format(
-                arg_label, target, quote_type_string(arg_type_str),
-                quote_type_string(expected_type_str))
+            if isinstance(outer_context, IndexExpr) and isinstance(outer_context.index, StrExpr):
+                msg = 'Argument "{}" has incompatible type {}; expected {}' .format(
+                    outer_context.index.value, quote_type_string(arg_type_str),
+                    quote_type_string(expected_type_str))
+            else:
+                msg = 'Argument {} {}has incompatible type {}; expected {}'.format(
+                    arg_label, target, quote_type_string(arg_type_str),
+                    quote_type_string(expected_type_str))
             code = codes.ARG_TYPE
             expected_type = get_proper_type(expected_type)
             if isinstance(expected_type, UnionType):

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -549,7 +549,7 @@ class MessageBuilder:
                     bare=True)
                 arg_label = '"{}"'.format(arg_name)
             if isinstance(outer_context, IndexExpr) and isinstance(outer_context.index, StrExpr):
-                msg = 'Argument "{}" has incompatible type {}; expected {}' .format(
+                msg = 'Value of "{}" has incompatible type {}; expected {}' .format(
                     outer_context.index.value, quote_type_string(arg_type_str),
                     quote_type_string(expected_type_str))
             else:

--- a/test-data/unit/check-typeddict.test
+++ b/test-data/unit/check-typeddict.test
@@ -730,7 +730,7 @@ p['x'] = 1
 from mypy_extensions import TypedDict
 TaggedPoint = TypedDict('TaggedPoint', {'type': str, 'x': int, 'y': int})
 p = TaggedPoint(type='2d', x=42, y=1337)
-p['x'] = 'y'  # E: Argument "x" has incompatible type "str"; expected "int"
+p['x'] = 'y'  # E: Value "x" has incompatible type "str"; expected "int"
 [builtins fixtures/dict.pyi]
 
 [case testCannotSetItemOfTypedDictWithInvalidStringLiteralKey]

--- a/test-data/unit/check-typeddict.test
+++ b/test-data/unit/check-typeddict.test
@@ -730,7 +730,7 @@ p['x'] = 1
 from mypy_extensions import TypedDict
 TaggedPoint = TypedDict('TaggedPoint', {'type': str, 'x': int, 'y': int})
 p = TaggedPoint(type='2d', x=42, y=1337)
-p['x'] = 'y'  # E: Argument 2 has incompatible type "str"; expected "int"
+p['x'] = 'y'  # E: Argument "x" has incompatible type "str"; expected "int"
 [builtins fixtures/dict.pyi]
 
 [case testCannotSetItemOfTypedDictWithInvalidStringLiteralKey]

--- a/test-data/unit/check-typeddict.test
+++ b/test-data/unit/check-typeddict.test
@@ -730,7 +730,7 @@ p['x'] = 1
 from mypy_extensions import TypedDict
 TaggedPoint = TypedDict('TaggedPoint', {'type': str, 'x': int, 'y': int})
 p = TaggedPoint(type='2d', x=42, y=1337)
-p['x'] = 'y'  # E: Value "x" has incompatible type "str"; expected "int"
+p['x'] = 'y'  # E: Value of "x" has incompatible type "str"; expected "int"
 [builtins fixtures/dict.pyi]
 
 [case testCannotSetItemOfTypedDictWithInvalidStringLiteralKey]

